bindings/node/test/providers/credentialsProvider.test.js

```diff
@@ -0,0 +1,527 @@
+          httpSpy.resolves(mockResponse);
+          await fetchAzureKMSToken({
+            url,
+            headers: {
+              customHeader1: 'value1',
+              customHeader2: 'value2'
+            }
+          });
+        });
+
+        it('allows a custom URL to be specified', () => {
+          const url = httpSpy.args[0][0];
+          expect(url).to.be.instanceof(URL);
+          expect(url.toString()).to.include('http://customentpoint.com');
+        });
+
+        it('deep copies the provided url', () => {
+          const spiedUrl = httpSpy.args[0][0];
+          expect(spiedUrl).to.be.instanceof(URL);
+          expect(spiedUrl).to.not.equal(url);
+        });
+
+        it('allows custom headers to be specified', () => {
+          const options = httpSpy.args[0][1];
+          expect(options).to.have.property('headers').to.have.property('customHeader1', 'value1');
+          expect(options).to.have.property('headers').to.have.property('customHeader2', 'value2');
+        });
+      });
+    });
+
+    context('error handling', () => {
+      afterEach(() => sinon.restore());
+      context('when the request times out', () => {
+        before(() => {
+          sinon.stub(utils, 'get').rejects(new MongoCryptKMSRequestNetworkTimeoutError());
+        });
+
+        it('throws a MongoCryptKMSRequestError', async () => {
+          const error = await loadCredentials({ azure: {} }).catch(e => e);
+          expect(error).to.be.instanceOf(MongoCryptAzureKMSRequestError);
+        });
+      });
+
+      context('when the request returns a non-200 error', () => {
+        context('when the request has no body', () => {
+          before(() => {
+            sinon.stub(utils, 'get').resolves({ status: 400 });
+          });
+
+          it('throws a MongoCryptKMSRequestError', async () => {
+            const error = await loadCredentials({ azure: {} }).catch(e => e);
+            expect(error).to.be.instanceOf(MongoCryptAzureKMSRequestError);
+            expect(error).to.match(/Malformed JSON body in GET request/);
+          });
+        });
+
+        context('when the request has a body', () => {
+          beforeEach(() => {
+            sinon
+              .stub(utils, 'get')
+              .resolves({ status: 400, body: '{ "error": "something went wrong" }' });
+          });
+
+          it('throws a MongoCryptKMSRequestError', async () => {
+            const error = await loadCredentials({ azure: {} }).catch(e => e);
+            expect(error).to.be.instanceOf(MongoCryptAzureKMSRequestError);
+          });
+
+          it('attaches the body to the error', async () => {
+            const error = await loadCredentials({ azure: {} }).catch(e => e);
+            expect(error).to.have.property('body').to.deep.equal({ error: 'something went wrong' });
+          });
+        });
+      });
+
+      context('when the request returns a 200 response', () => {
+        context('when the request has no body', () => {
```

Comment:
might want to add a case for when the body is defined but not json

Reply:
added!